common/io.cpp

```diff
@@ -0,0 +1,62 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <common/io.hpp>
+
+namespace mender::common::io {
+
+StdError Copy(Writer &dst, Reader &src) {
+	vector<uint8_t> buffer(4096);
+	return Copy(dst, src, buffer);
+}
+
+StdError Copy(Writer &dst, Reader &src, vector<uint8_t> &buffer) {
+	size_t orig_size = buffer.size();
+
+	while (true) {
+		if (buffer.size() != orig_size) {
+			buffer.resize(orig_size);
+		}
+
+		auto result = src.Read(buffer);
+		if (!result) {
+			return result.error();
+		} else if (result.value() == 0) {
+			return StdError();
+		} else if (result.value() > buffer.size()) {
+			return StdError(
+				std::error_condition(std::errc::invalid_argument),
```

Comment:
I don't think this is `invalid_argument`.

Reply:
I wanted to separate it from `io_error` because ending up here is guaranteed to be a bug, whereas the other `io_error` returns can happen as a result of real I/O conditions. Do you have a better suggestion from [this list](https://en.cppreference.com/w/cpp/error/errc)?

Another alternative could be to use our own error category (which we just talked about) and define it there, but it seems a bit overkill for something that isn't supposed to happen in the first place.

Reply:
> something that isn't supposed to happen in the first place.

I would add `assert(false)` then, in the first place. :)

Reply:
> Do you have a better suggestion from [this list](https://en.cppreference.com/w/cpp/error/errc)?

I don't think we should limit ourselves to any list.

Reply:
Switching to `error_condition` luckily solves this problem as well, because now we can use enums from any category and mix them. So I'll add an error code for programming error.

Reply:
(I forgot about this one. It's coming!)

Reply:
Ok, now it's fixed.